solrmonitor/zkdispatcher.go

```diff
@@ -24,24 +24,49 @@ import (
 	"github.com/samuel/go-zookeeper/zk"
 )
 
-// Handle a zk event.  Optionally return a new event channel to automatically re-register.
+// ugh... seems silly to have to define this
+const maxInt = int(^uint(0) >> 1)
+
+// A function that handles a zk event. Optionally returns a new event channel to automatically
+// re-register.
 type ZkEventHandler func(zk.Event) <-chan zk.Event
 
-// Monitors many zk.Event channels on a single goroutine.
+// An interface that handles a zk event. Optionally returns a new event channel to automatically
+// re-register.
+type ZkEventCallback interface {
+	Handle(zk.Event) <-chan zk.Event
+}
+
+// Adapts a ZkEventHandler function to the ZkEventCallback interface.
+type zkEventHandlerAdapter struct {
+	handler ZkEventHandler
+}
+
+func (a *zkEventHandlerAdapter) Handle(event zk.Event) <-chan zk.Event {
+	return a.handler(event)
+}
+
+// Monitors many zk.Event channels. Dispatches handling to other goroutines allowing handlers to
+// run in parallel. A single ZkEventCallback, however, will always be invoked as if from the
+// same goroutine. So a single callback can be used to watch multiple channels and it will receive
+// events in a way that does not require it to synchronize with other possible invocations.
```

Comment:
The simplest thing would be to just spin off a goroutine for every single dispatched event. 

Another approach I tried was similar to what you suggested -- a fixed number of goroutines polling queues (though hashing watchers to a given queue, so it still provided thread confinement). While relaxing the ordering provision could simplify the code, I think it makes for a harder-to-use programming model and could complicate every usage of the API -- especially if a single watcher were used to monitor an entire sub-tree that could have a very large number of nodes (a la something like tree cache).

Aside from the question on ordering, there is a problem in having to pick the number of watchers. Fixing the maximum number of goroutines could just reduce throughput -- like cases where there are enough ready listeners to saturate CPU, but they are stuck in the queue behind listeners that are blocking.

With the approach in this branch, the number of goroutines is limited to the number of distinct handlers registered -- and even then it could only get that high if every single handler had a task an event to process simultaneously. So it should be about optimal for throughput while still having a reasonable limit on goroutines (at worst, still better than the approach where every watch channel has a dedicated goroutine).